File: Ejercicio_3/fibonacci.py
def fibonacci(n):
    def op(a, b, c):
        return b, c, a+b+c

    def combinar(initial, op, n):
        a, b, c = initial
        for i in range(n):
            a, b, c = op(a, b, c)
        return b

    if n == 0:
        return 0
    if n == 1:
        return 1
    if n == 2:
        return 1

    return combinar((0, 1, 1), op, n - 1)


print("Fibonacci con funciones de orden Superior")
n = int(input("Introduzca un numero: "))
for i in range(n):
    print(fibonacci(i), end=" ")

File: Ejercicio_3/fibonacci.scala
object Fibonacci{

  def FiboRecursivo(n: Int): Int = {
    if (n == 0) 0
    else if (n == 1) 0
    else if (n == 2) 1
    else FiboRecursivo(n-1) + FiboRecursivo(n-2)+FiboRecursivo(n-3)
  }

  def fsup(ftribR:(Int)=>Int):Unit={
    print("Introduzca un numero: ")
    val n=scala.io.StdIn.readInt()
    for (i <- 0 to n){
        print(ftribR(i) + " ")
    }
  }

  def main(args:Array[String]):Unit={    
    fsup(FiboRecursivo)
  } 

}